src/lib/coda_transition/external_transition.ml

```diff
@@ -657,6 +792,103 @@ struct
         ~error:`Parent_missing_from_frontier
     in
     (t, Validation.Unsafe.set_valid_frontier_dependencies validation)
+
+  let validate_delta_transition_chain_part2 (t, validation) ~frontier =
+    let open Result.Let_syntax in
+    let open Int in
+    let global_slot =
+      t |> With_hash.data |> consensus_state
+      |> Consensus.Data.Consensus_state.global_slot
+    in
+    let boundary_hash =
+      t |> With_hash.data |> delta_transition_chain_proof |> fst
+    in
+    if State_hash.equal boundary_hash (With_hash.hash t) then
+      let out_of_boundary_hash = t |> With_hash.data |> parent_hash in
+      match
+        Transition_frontier.find frontier out_of_boundary_hash
+        |> Option.map ~f:Transition_frontier.Breadcrumb.validated_transition
+      with
+      | None ->
+          Ok
+            ( t
+            , Validation.Unsafe.set_valid_delta_transition_chain_part2
+                validation )
+      | Some out_of_boundary_transition ->
+          let out_of_boundary_global_slot =
+            out_of_boundary_transition |> Validated.consensus_state
+            |> Consensus.Data.Consensus_state.global_slot
+          in
+          let%map () =
+            Result.ok_if_true
+              ( out_of_boundary_global_slot
+              < global_slot - Consensus.Constants.delta )
+              ~error:`Invalid_delta_transition_chain_proof
+          in
+          ( t
+          , Validation.Unsafe.set_valid_delta_transition_chain_part2 validation
+          )
+    else
+      match
+        Transition_frontier.find frontier boundary_hash
+        |> Option.map ~f:Transition_frontier.Breadcrumb.validated_transition
+      with
+      | None ->
+          Ok
+            ( t
+            , Validation.Unsafe.set_valid_delta_transition_chain_part2
+                validation )
+      | Some boundary_transition -> (
+          let boundary_slot =
+            Validated.consensus_state boundary_transition
+            |> Consensus.Data.Consensus_state.global_slot
+          in
+          let%bind () =
+            Result.ok_if_true
+              (boundary_slot > global_slot - Consensus.Constants.delta)
```

Comment:
It would be good to have more specific errors for each of these cases of validation that fails. Do the logs otherwise contain enough information to precisely determine why this function fails, if it fails?